files/spark/bin/charmed-spark-entrypoint.sh

```diff
@@ -1,9 +1,24 @@
 #!/bin/bash
 
-# This script still exists because passing of working directory to services has not yet landed
-# in Pebble (https://github.com/canonical/pebble/issues/158) and being in the write directory
-# is important for the entrypoint script
+TYPE=$1
 
-cd /opt/spark
+echo "Running script with ${TYPE} flavour"
 
-./entrypoint.sh $*
\ No newline at end of file
+case "${TYPE}" in
```

Comment:
This comment is related to the one below, as I think there is a mixup of services and entrypoints.

The use of this script kinda competes with the purpose of Pebble and the management of multiple processes.

Please consider the following proposal:

**Dockerfile**
```dockerfile
# Set the default dir
WORKDIR /opt/spark
# leave the service name out
ENTRYPOINT [ "/bin/pebble", "enter", "--verbose", "--args" ]
```
---

**files/spark/bin/charmed-spark-entrypoint.sh**

remove this file

---

**rockcraft.yaml**
```yaml
services:
  driver:
    command: /opt/spark/entrypoint.sh driver [ foo ]
...
  executor:
    command: /opt/spark/entrypoint.sh executor [ foo ]
...
  history-server:
    command: /opt/spark/sbin/start-history-server.sh --properties-file /etc/spark/conf/spark-defaults.conf
```

Now after writing all this, I realize that your problem could be related with the fact that you want these separate commands to be disabled by default, but then you can't use the pebble client command `start <svc>`. If that's the case, then I think your solution might be the only one. Nonetheless, what is presented here gives you, at least, a few optimization points:
 1. you can just set `WORKDIR` to avoid `cd /opt/spark` in the script
 2. you can delete the `history-server` service cause you don't seem to ever use it (?)

Reply:
Yes. I've actually been thinking a way to keep the entrypoint service disabled (since I believe it would be cleaner), but one of the problems is exactly that I don't have a way to then enable the service while keeping the same UX for launching the image. 

A though I had (but I have not tested this and it sounded to me that I was complicating things a bit too much) was to have an entrypoint (still enabled by default) that basically start pebble services via Pebble API (`pebble start <argument>`).

Reply:
1. => Yes, that's fine. But the problem is that everything we put in Dockerfile should be then provided by rockcraft. If this feature can land in rockcraft soon (such that we are then able to by-pass the temp workaround of using Dockerfile), I'm more than fine to change that

2. => Yes, it is not very critical or important to have the `history-server`. I just though that it would have been nice to allow a clean UX for running the image with docker as well, e.g. `docker run charmed-spark history-server`, because as it stands now one should do: `docker run charmed-spark \; start history-server`, with the `\;` which is somewhat misleading and "strange". Also (but this is a small point), in the second case you would still run the bash script with an infinite loop there. I don't have strong position on this, and I'd also like to hear @grobbie thoughts, since it is a bit of a product UX decision whether we want to allow the `docker run charmed-spark history-server` clean UX

Reply:
We should not break the existing UX of the Spark container for users that don't choose to use the charm. So we need to support `docker run charmed-spark history-server` (as well as thrift-gateway)

Reply:
I have talked to Rob about 2). We may want to provide this UX, but probably we should address this as a separate ticket, as this can be seen as a new features we'd like to provide, and we'd better think this through (and implement this consistently for different functionalities we want to expose)

Reply:
@cjdcordeiro What about (1)? Is there any plan to land the two features (default entrypoint service and working dir) in rockcraft soon?

Reply:
> But the problem is that everything we put in Dockerfile should be then provided by rockcraft.

well, I wouldn't be too sure that the fiddling of the entrypoint would be something that will ever be supported by Rockcraft.

When it comes to the working-directory though, that's a planned feature for pebble: https://github.com/canonical/pebble/issues/158 (looks kinda stale, but I've heard that `contexts` are being revived so this might go along as well :crossed_fingers: )

Reply:
Yes yes, I meant that issue. In the end, we need to set the working directory at pebble service level. Since it has been stable, I'd just keep the things that Docker does to a bare minimun, such that we can migrate to getting rid of that asap. So, to me, providing a default entrypoint serivce is definitely higher priority, since we cannot go ahead otherwise

Reply:
> So, to me, providing a default entrypoint serivce is definitely higher priority, since we cannot go ahead otherwise

that's the real question mark. The reason why `primary` services never came to be was precisely that it was impossible to harmonize the Pebble CLI with the services' arguments. Allowing a mechanism where this

```
docker run <rock1> --args svcName --foo bar
```

is as valid as 

```
docker run <rock2> --foo bar
```

will go against the ROCK's main principle of "offering a consistent interface for all ROCKs". Changing this means changing some of the fundamental principles of ROCKs. I'm open to discussing it though...there might be a brilliant idea out there that we haven't considered yet :)